core/platform/lf_nrf52_support.c

```diff
@@ -0,0 +1,310 @@
+        _lf_sleep_interrupted = false;
+    } else if (event_type == NRF_TIMER_EVENT_COMPARE3) {
+        _lf_time_us_high =+ 1;
+    }
+}
+
+/**
+ * Initialize the LF clock.
+ */
+void lf_initialize_clock() {
+    ret_code_t error_code;
+    _lf_time_us_high = 0;
+
+    // Initialize TIMER3 as a free running timer
+    // 1) Set to be a 32 bit timer
+    // 2) Set to count at 1MHz
+    // 3) Clear the timer
+    // 4) Start the timer
+
+    nrfx_timer_config_t timer_conf = {
+        .frequency = NRF_TIMER_FREQ_1MHz,
+        .mode = NRF_TIMER_MODE_TIMER,
+        .bit_width = NRF_TIMER_BIT_WIDTH_32,
+        .interrupt_priority = 7, // lowest
+        .p_context = NULL,
+    };
+
+    error_code = nrfx_timer_init(&g_lf_timer_inst, &timer_conf, &lf_timer_event_handler);
+    APP_ERROR_CHECK(error_code);
+    // Enable an interrupt to occur on channel NRF_TIMER_CC_CHANNEL3
+    // when the timer reaches its maximum value and is about to overflow.
+    nrfx_timer_compare(&g_lf_timer_inst, NRF_TIMER_CC_CHANNEL3, 0x0, true);
+    nrfx_timer_enable(&g_lf_timer_inst);
+}
+
+/**
+ * Fetch the value of _LF_CLOCK (see lf_linux_support.h) and store it in tp. The
+ * timestamp value in 't' will will be the number of nanoseconds since the board was reset.
+ * The timers on the board have only 32 bits and their resolution is in microseconds, so
+ * the time returned will always be an even number of microseconds. Moreover, after about 71
+ * minutes of operation, the timer overflows. 
+ * 
+ * The function reads out the upper word before and after reading the timer.
+ * If the upper word has changed (i.e. it was an overflow in between),
+ * we cannot simply combine them. We read once more to be sure that 
+ * we read after the overflow.
+ *
+ * @return 0 for success, or -1 for failure. In case of failure, errno will be
+ *  set appropriately (see `man 2 clock_gettime`).
+ */
+int lf_clock_gettime(instant_t* t) {
+    if (t == NULL) {
+        // The t argument address references invalid memory
+        // errno = EFAULT; //TODO: why does this not work with new build process?
+        return -1;
+    }
+    uint32_t now_us_hi_pre = _lf_time_us_high;
+    uint32_t now_us_low = nrfx_timer_capture(&g_lf_timer_inst, NRF_TIMER_CC_CHANNEL1);
+    uint32_t now_us_hi_post = _lf_time_us_high; 
+
+    // Check if we read the time during a wrap
+    if (now_us_hi_pre != now_us_hi_post) {
+        // There was a wrap. read again and return
+        now_us_low = nrfx_timer_capture(&g_lf_timer_inst, NRF_TIMER_CC_CHANNEL1);
+    }
+    uint64_t now_us = COMBINE_HI_LO(now_us_hi_post, now_us_low);
+
+    *t = ((instant_t)now_us) * 1000;
+
+    return 0;
+}
+
+/**
+ * @brief Return whether the critical section has been entered.
+ * 
+ * @return true if interrupts are currently disabled
+ * @return false if interrupts are currently enabled
+ */
+bool in_critical_section() {
+    // FIXME: if somehow interrupts get disabled directly (not through the NRF API),
```

Comment:
See my suggested implementation in the Arduino code for an alternative that will address this FIXME and also support nested critical sections.